core/src/main/java/com/google/android/fhir/sync/FhirSynchronizer.kt

```diff
@@ -49,4 +54,63 @@ class FhirSynchronizer(
       return Result.Error(exceptions)
     }
   }
+
+  suspend fun upload(): Result {
+    val exceptions = mutableListOf<ResourceSyncException>()
+    database.getAllLocalChanges().forEach {
+      try {
+        val response: Resource = doUpload(it.localChange, dataSource)
+        if (response.logicalId.equals(it.localChange.resourceId) || response.isSuccess()) {
+          database.deleteUpdates(it.token)
+        } else {
+          // TODO improve exception message
+          exceptions.add(
+            ResourceSyncException(
+              ResourceType.valueOf(it.localChange.resourceType),
+              FHIRException(
+                "Could not infer response \"${response.resourceType}/${response.logicalId}\" as success."
+              )
+            )
+          )
+        }
+      } catch (exception: Exception) {
+        exceptions.add(
+          ResourceSyncException(ResourceType.valueOf(it.localChange.resourceType), exception)
+        )
+      }
+    }
+
+    if (exceptions.isEmpty()) {
+      return Result.Success
+    } else {
+      return Result.Error(exceptions)
+    }
+  }
+
+  private suspend fun doUpload(
+    localChange: LocalChangeEntity,
+    datasource: FhirDataSource
+  ): Resource =
+    when (localChange.type) {
+      LocalChangeEntity.Type.INSERT ->
+        datasource.insert(localChange.resourceType, localChange.resourceId, localChange.payload)
+      LocalChangeEntity.Type.UPDATE ->
+        datasource.update(localChange.resourceType, localChange.resourceId, localChange.payload)
+      LocalChangeEntity.Type.DELETE ->
+        datasource.delete(localChange.resourceType, localChange.resourceId)
+    }
+
+  /**
+   * Determines if the upload operation was successful or not.
+   *
+   * Current HAPI FHIR implementation does not give any signal other than 'severity' level for
+   * operation success/failure. TODO: pass along the HTTP result (or any other signal) to determine
+   * the outcome of an instance level RESTful operation.
+   */
+  private fun Resource.isSuccess(): Boolean {
+    if (!this.resourceType.equals(ResourceType.OperationOutcome)) return false
+    val outcome: OperationOutcome = this as OperationOutcome
+    return outcome.hasIssue() &&
```

Comment:
should this be

```suggestion
    return !outcome.hasIssue() ||
```

?

Reply:
I am using `Resource.hasXX()` to guard against empty/null properties. Maybe `outcome.issue?.all... ?: false` is better?

Reply:
so if `outcome` has no `issue`, would you return `true` or `false`? I kinda feel it should be `true`? But you would return `false` here?

Reply:
I was going with `false` because it does not imply success or failure. And I think better to fail fast. That being said we need to have robust signals for operation success. I have created #440 to track it.